impl/sql/mutationstorage/queue.go

```diff
@@ -34,23 +37,67 @@ func (m *Mutations) Send(ctx context.Context, domainID string, update *pb.EntryU
 	if err != nil {
 		return err
 	}
-	writeStmt, err := m.db.Prepare(insertQueueExpr)
+	return m.send(ctx, domainID, mData, time.Now())
+}
+
+// ts must be greater than all other timestamps currently recorded for domainID.
+func (m *Mutations) send(ctx context.Context, domainID string, mData []byte, ts time.Time) error {
+	tx, err := m.db.BeginTx(ctx, &sql.TxOptions{Isolation: sql.LevelSerializable})
 	if err != nil {
 		return err
 	}
-	defer writeStmt.Close()
-	_, err = writeStmt.ExecContext(ctx, domainID, time.Now().UnixNano(), mData)
-	return err
+
+	var maxTime int64
+	if err := tx.QueryRowContext(ctx,
+		`SELECT COALESCE(MAX(Time), 0) FROM Queue WHERE DomainID = ?;`,
+		domainID).Scan(&maxTime); err != nil {
+		if rollbackErr := tx.Rollback(); rollbackErr != nil {
+			return status.Errorf(codes.Internal, "could not roll back: %v\n", rollbackErr)
+		}
+		return err
+	}
+	tsTime := ts.UnixNano()
+	if tsTime <= maxTime {
+		if rollbackErr := tx.Rollback(); rollbackErr != nil {
+			return status.Errorf(codes.Internal, "could not roll back: %v\n", rollbackErr)
+		}
+		return status.Errorf(codes.Aborted, "timestamp: %v, want > %v", tsTime, maxTime)
```

Comment:
This error message could be clearer, to help out anyone who's in the position of having to investigate this.  Maybe something like "max-timestamp %v of queued mutations > current time %v"?